GAT + enum: incorrect unused lifetime error

This is a follow up on #67089, which now works. The following doesn't ([play link](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=d6570106bebc880bc301f30de216b611)):
```rust
#![feature(generic_associated_types)]

trait A {
    type B<'a>;
    
    fn make_b<'a>(&'a self) -> Self::B<'a>;
}

struct S {}

impl A for S {
    type B<'a> = &'a S;
    fn make_b<'a>(&'a self) -> &'a Self {
        self
    }
}

enum E<'a> {
    S(<S as A>::B<'a>),
}
```
Error is:
```
18 | enum E<'a> {
   |        ^^ unused parameter
```
however, this lifetime clearly is used and is required.

On this topic, the following builds,
```rust
trait Foo {}
impl<'a, T: A> Foo for T
where <T as A>::B<'a>: std::fmt::Debug
{}
```
but the bound `<T as A>::B<'_>: std::fmt::Debug` fails with:
```
error[E0637]: `'_` cannot be used here

I *think* this should work?
```

Also, the following fails:
```rust
trait Foo {}
impl<'a, T: A> Foo for T
where <T as A>::B<'a>: 'static
{}
```
with:
```
error[E0310]: the associated type `<T as A>::B<'_>` may not live long enough
```
This bound may not make much sense (since there is another lifetime parameter), but I believe it should be valid to specify here.

--- a/compiler/rustc_typeck/src/variance/constraints.rs
+++ b/compiler/rustc_typeck/src/variance/constraints.rs
@@ -207,27 +207,13 @@ impl<'a, 'tcx> ConstraintContext<'a, 'tcx> {
         }
     }
 
-    fn add_constraints_from_trait_ref(
-        &mut self,
-        current: &CurrentItem,
-        trait_ref: ty::TraitRef<'tcx>,
-        variance: VarianceTermPtr<'a>,
-    ) {
-        debug!("add_constraints_from_trait_ref: trait_ref={:?} variance={:?}", trait_ref, variance);
-        self.add_constraints_from_invariant_substs(current, trait_ref.substs, variance);
-    }
-
+    #[instrument(skip(self, current))]
     fn add_constraints_from_invariant_substs(
         &mut self,
         current: &CurrentItem,
         substs: SubstsRef<'tcx>,
         variance: VarianceTermPtr<'a>,
     ) {
-        debug!(
-            "add_constraints_from_invariant_substs: substs={:?} variance={:?}",
-            substs, variance
-        );
-
         // Trait are always invariant so we can take advantage of that.
         let variance_i = self.invariant(variance);
 
@@ -300,8 +286,7 @@ impl<'a, 'tcx> ConstraintContext<'a, 'tcx> {
             }
 
             ty::Projection(ref data) => {
-                let tcx = self.tcx();
-                self.add_constraints_from_trait_ref(current, data.trait_ref(tcx), variance);
+                self.add_constraints_from_invariant_substs(current, data.substs, variance);
             }
 
             ty::Opaque(_, substs) => {

--- a/src/test/ui/generic-associated-types/variance_constraints.rs
+++ b/src/test/ui/generic-associated-types/variance_constraints.rs
@@ -1,0 +1,24 @@
+// check-pass
+// issue #69184
+#![feature(generic_associated_types)]
+#![allow(incomplete_features)]
+
+trait A {
+    type B<'a>;
+
+    fn make_b<'a>(&'a self) -> Self::B<'a>;
+}
+
+struct S {}
+impl A for S {
+    type B<'a> = &'a S;
+    fn make_b<'a>(&'a self) -> &'a Self {
+        self
+    }
+}
+
+enum E<'a> {
+    S(<S as A>::B<'a>),
+}
+
+fn main() {}

